Replace instance from the team library component

rei-suzuki opened the pull request:
Signed-off-by: rei-suzuki <tcoxoxol8@gmail.com>

rei-suzuki committed:
Replace instance from the team library component

Signed-off-by: rei-suzuki <tcoxoxol8@gmail.com>

rei-suzuki committed:
Get components from team library with set of layer name and key

Signed-off-by: rei-suzuki <tcoxoxol8@gmail.com>

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,53 @@
+    await replaceNodes(figma.currentPage.selection)
   }
-  figma.closePlugin()
+    figma.closePlugin()
```
```suggestion
  figma.closePlugin()
```

rei-suzuki replied:
直した https://github.com/rei-suzuki/figma-team-component-transporter/pull/1/commits/5f638c3305fe5429c6a94a6156d744f4a20b1486

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,53 @@
 }
 
 // ComponentFileのComponent IDを取得して保存
+const teamLibraryMasterComponents= []
```
L16にもっていこう

rei-suzuki replied:
直した https://github.com/rei-suzuki/figma-team-component-transporter/pull/1/commits/5f638c3305fe5429c6a94a6156d744f4a20b1486

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,53 @@
+const teamLibraryMasterComponents= []
 async function saveTargetComponent() {
-  return 1
+  for (let i = 0; i< figma.currentPage.selection.length; i++) {
+    let nodes = figma.currentPage.selection
+
+    teamLibraryMasterComponents.push(new MasterComponent(nodes[i].name, (nodes[i] as ComponentNode).key))
+    await figma.clientStorage.setAsync(clientStrageKey, teamLibraryMasterComponents)
+    }
+    
+  return true
```
- selectionごとに保存しなくても、最終的にできたものだけ保存すればいい（ループの外に出せる）
- `MasterComponent` クラスだと、L32のように毎回findしないといけなくて計算量が増える
- nameは一意なので、 `{ key: value }` の形で保存すると一発で引いてこられる


```ts
async function saveTargetComponent(nodes) { // ついでにreplaceNodesの呼び出しと形式を揃えた
  teamLibraryMasterComponents = {}
  for (let i = 0; i< nodes.length; i++) {
    teamLibraryMasterComponents[nodes[i].name] = (nodes[i] as ComponentNode).key
  }
  await figma.clientStorage.setAsync(clientStrageKey, teamLibraryMasterComponents)
}
```

rei-suzuki replied:
https://github.com/rei-suzuki/figma-team-component-transporter/pull/1/commits/5f638c3305fe5429c6a94a6156d744f4a20b1486

- PageNodeからComponentとComponentSetだけreplaceの対象とした
- オブジェクトに変更に伴いreplaceNode側も諸々対応した

rei-suzuki committed:
Change name and key to objects and get the component from PageNode	1a7d31b

rei-suzuki committed:
Add node modules

rei-suzuki committed:
Replace the target node by recursive

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,65 @@
+    if (node.children != null) {
+      findComponent(teamLibraryMasterComponents, node.children)
+    }
+    if (node.children === null){
+      return false
+    }
```
これは削ってよさそう（ついでにreturnはループごと抜けちゃうので処理漏れが生まれそう）

rei-suzuki replied:
削除　https://github.com/rei-suzuki/figma-team-component-transporter/pull/1/commits/0a074618883a1d127eff3cf698704064ddf64106

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,65 @@
+async function findComponent(teamLibraryMasterComponents, nodes) {
+  const targetNode = []
+  for (const node of nodes) {
+    if (node.type === "COMPONENT" || node.type === "COMPONENT_SET") {
```
component setはcreateInstanceできないから判定から外しちゃってもいいんじゃないかな？

rei-suzuki replied:
確かに! https://github.com/rei-suzuki/figma-team-component-transporter/pull/1/commits/2635d09e69a61cb7ec1592c954e1d5fe5f5b4599

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,65 @@
-async function replaceNodes() {
-  return 1
+async function replaceNodes(nodes, teamLibraryComponents) {
+  for (let i = 0; i < nodes.length; i++) {
```
imo 上が `for (const node of nodes)` なのでこっちもそれに合わせたら可読性上がりそう

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,65 @@
+      }
+    } else if (nodes[i].children != null) {
+      replaceNodes(nodes[i].children, teamLibraryComponents)
+    } else if (nodes[i].children === null) {
+      return false
+    }
```
これも不要そう

rei-suzuki replied:
削除 https://github.com/rei-suzuki/figma-team-component-transporter/pull/1/commits/0a074618883a1d127eff3cf698704064ddf64106

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,65 @@
+      } else {
+        replaceNodes(nodes[i].children, teamLibraryComponents)
+      }
+    } else if (nodes[i].children != null) {
```
```suggestion
    }
    if (nodes[i].children != null) {
```

instance(frame) nodeかつchildrenがある場合にここを通らないので、別のif文にするのがいいね

rei-suzuki committed:
Remove component_set type

rei-suzuki committed:
Aligned write array loops

rei-suzuki committed:
Update code.ts

Remove code

rei-suzuki committed:
Update code.js

Add js

rei-suzuki committed:
Separate replace and find function

rei-suzuki committed:
Add gitignore

ar-tama committed:
Remove node_modules

ar-tama committed:
fixup! Remove node_modules

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,65 @@
+      if (key != undefined) { 
+        if (node.type === "INSTANCE" && node.mainComponent.key === teamLibraryComponents[node.name]) {
+          continue
+        } else {
+          try {  
```
- continueでループを抜けるのでL40のelseは不要
- try catchの処理もreplaceNodesに寄せよう

rei-suzuki replied:
なるほど...!!! https://github.com/rei-suzuki/figma-team-component-transporter/pull/1/commits/089f24f18c8e57ba3b0d226fb132ebe90545de74
L9の呼び出す関数を直すのミスってたので修正

rei-suzuki committed:
Fix review comment

rei-suzuki commented on code.ts:
```diff
@@ -53,13 +47,19 @@ async function findTargetNodes(nodes, teamLibraryComponents) {
+    node.remove()
+  }
+  catch(e) {
+    const error = e.toString()
```
エラーがオブジェクトできてるよって怒られたので追加した

rei-suzuki committed:
Fix name of menu and function

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,65 @@
+
+async function replaceNodes(node, key) {
+  try {  
+    const getTeamLibraryComponent = await figma.importComponentByKeyAsync(key)
```
```suggestion
    const teamLibraryComponent = await figma.importComponentByKeyAsync(key)
```

getTeamLibraryComponent だとメソッド名みたいに見えてしまうかな〜

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,65 @@
+  }
+}
+
+async function replaceNodes(node, key) {
```
```suggestion
async function replaceNode(node, key) {
```

対象が単数なのでメソッドも単数に

ar-tama commented on code.ts:
```diff
@@ -1,21 +1,65 @@
 // 保存したComponent IDを使ってリプレイス
-async function replaceNodes() {
-  return 1
+async function findTargetNodes(nodes, teamLibraryComponents) {
```
```suggestion
async function scanNodes(nodes, teamLibraryComponents) {
```

- findだと「見つけたものを返す」という振る舞いが想起されるので、変えたい
- 意味としては `scan and replace` みたいなものが適当そう
  - `scanAndReplaceNodes` でもいいけど
  - `scanNodes` でも順当かなという気がする（imo）
- ちなみにユーザからの見え方は変わらず `replace` もしくは `transport` なんかでいいと思う（コマンド名）
  - 再帰的に走査して置換してまわるのは内部の話なので

rei-suzuki committed:
Fix variable name

https://github.com/rei-suzuki/figma-team-component-transporter/pull/1#discussion_r585211355

Co-authored-by: Makoto Arata <arata.makoto@gmail.com>

rei-suzuki committed:
Fix variable names to singular

Co-authored-by: Makoto Arata <arata.makoto@gmail.com>

rei-suzuki committed:
Fix variable names to plural

rei-suzuki committed:
Merge branch 'get_component_replace_node' of https://github.com/rei-suzuki/figma-team-component-transporter into get_component_replace_node

rei-suzuki committed:
Fix find function name

rei-suzuki committed:
Fix replace menu name

rei-suzuki committed:
Fix plugin name

rei-suzuki committed:
Add description

ar-tama reviewed (approved):
🙌
